File: 3. Behavioral Design Patterns/Strategy/Python/StrategyTemplate.py
# Strategy Pattern – Python Template

from abc import ABC, abstractmethod

class Strategy(ABC):
    @abstractmethod
    def execute(self, data):
        pass

class ConcreteStrategyA(Strategy):
    def execute(self, data):
        return f"Strategy A: {data}"

class ConcreteStrategyB(Strategy):
    def execute(self, data):
        return f"Strategy B: {data}"

class Context:
    def __init__(self, strategy):
        self.strategy = strategy
    
    def set_strategy(self, strategy):
        self.strategy = strategy
    
    def do_work(self, data):
        return self.strategy.execute(data)

def main():
    context = Context(ConcreteStrategyA())
    print(context.do_work("test"))
    context.set_strategy(ConcreteStrategyB())
    print(context.do_work("test"))

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Strategy/Java/StrategyTemplate.java
// Strategy Pattern – Java Template

import java.util.*;

interface Strategy { List<Integer> execute(List<Integer> data); }
class ConcreteStrategyA implements Strategy { public List<Integer> execute(List<Integer> data){ return data; } }
class ConcreteStrategyB implements Strategy { public List<Integer> execute(List<Integer> data){ return data; } }
class Context { private Strategy s; Context(Strategy s){ this.s = s; } void set(Strategy s){ this.s = s; } List<Integer> doWork(List<Integer> data){ return s.execute(data); } }
public class StrategyTemplate { public static void main(String[] args){ new Context(new ConcreteStrategyA()).doWork(new ArrayList<>()); } }
